config/derive/src/proxy.rs

```diff
@@ -0,0 +1,212 @@
+use proc_macro2::TokenStream;
+use proc_macro_error::{abort, OptionExt};
+use quote::{format_ident, quote};
+use syn::{
+    parse_quote,
+    visit_mut::{visit_derive_input_mut, VisitMut},
+};
+
+pub fn impl_builder(mut builder_struct: syn::DeriveInput) -> TokenStream {
+    let parent_ident = builder_struct.ident.clone();
+
+    let mut builder_resolver = BuilderResolver::new();
+    builder_resolver.visit_derive_input_mut(&mut builder_struct);
+
+    let proxy_struct_name = &builder_struct.ident;
+
+    let set_field_methods = builder_resolver.set_field_methods;
+    let field_names = builder_resolver.field_names;
+
+    let impl_parent_default = if !builder_resolver.has_required_field {
+        let ident = field_names
+            .iter()
+            .map(|field_value| format_ident!("DEFAULT_{}", field_value.to_string().to_uppercase()));
+
+        quote! {
+            impl Default for #parent_ident {
+                fn default() -> Self {
+                    Configuration::new()
+                }
+            }
+
+            impl #parent_ident {
+                pub const fn new() -> Self {
+                    Self { #( #field_names: Self::#ident ),* }
+                }
+            }
+        }
+    } else {
+        quote!()
+    };
```

Comment:
#### `#[config(required)]`

It seems that `#[config(required)]` works like "if there is no value provided, fallback to a default value". For me it wasn't semantically obvious.

First of all, I think this fallback should not happen in `ConfigurationBuilder`. Multiple `ConfigurationBuidler`s are merged anyway, so you set defaults multiple times. It seems more logically to have ALL fields optional in `ConfigurationBuilder` (like it was in `ConfigurationProxy`). However, when `ConfigurationBuilder` transforms into `Configuration` - this is the right place to put defaults instead of the missing fields. 

Also, giving `#[config(required)]` attribute to a field means that the final `Configuration` must have this field defined. Since most of the fields must be defined anyway, it seems more logically to mark rest optional fields with `#[config(optional)]` instead of marking the most required fields with `#[config(required)]`.

And also, there is no need to use `Option<Option<..>>` type in `ConfigurationBuilder` fields. All of the fields might be simple `Option<..>`, but if the field is marked as `#[config(optional)]`, it is also `Option<..>` in the `Configuration`. it changes behaviour during the configuration build (i.e. `ConfigurationBuilder` → `Configuration`): 

```rust
#[derive(Configuration)]
struct Config {
  foo: String,
  #[config(optional)]
  bar: u32
}

// produced code:

struct ConfigBuilder {
  foo: Option<String>,
  bar: Option<u32>
}

struct Config {
  foo: String,
  bar: Option<u32>
}
```

Does it make sense for you?

#### Default values

Could we implement defaults binding in the same way `serde` does it?

```rust
struct Config {
  // inline
  #[config(default_value_t = "512")]
  limit: u32

  // function
  #[config(default_value = "get_default_log_level")]
  log_level: LogLevel
}

fn get_default_log_level() -> LogLevel {
  LogLevel::INFO
}
```

Macro can also auto-generate associated consts:

```rust
trait Config {
  const DEFAULT_LIMIT: u32 = 512;
  
  // if fn is const
  const DEFAULT_LOG_LEVEL: LogLevel = get_default_log_level();
  
  // if fn is not const
  fn DEFAULT_LOG_LEVEL() -> LogLevel {
    get_default_log_level()
  }
}
```

Reply:
> It seems that #[config(required)] works like "if there is no value provided, fallback to a default value". For me it wasn't semantically obvious.

this is not how `config(required)` works. If the field is `config(required)` it doesn't have a default

> First of all, I think this fallback should not happen in ConfigurationBuilder. Multiple ConfigurationBuidlers are merged anyway, so you set defaults multiple times. It seems more logically to have ALL fields optional in ConfigurationBuilder (like it was in ConfigurationProxy). However, when ConfigurationBuilder transforms into Configuration - this is the right place to put defaults instead of the missing fields.

I think this is exactly how it works right now. There are 3 possible sources for `ConfigurationBuilder`, the `Default::default`, reading from file and reading from environment. Additionally, there is `override_with` method on the builder so you can combine various sources into one configuration 

> Also, giving #[config(required)] attribute to a field means that the final Configuration must have this field defined

`config(required)` means that it must be given by the user. This I should make more clear in the docs

> Could we implement defaults binding in the same way serde does it?

we could, but I maintain that default values should be constant so it doesn't make sense to make it expression

Reply:
when I started I thought that for config like this:
```rs
#[derive(Configuration)]
struct Config {
  foo: String,
  #[config(required)]
  bar: u32
}
```

builder would look like:
```rs
#[derive(Configuration)]
struct ConfigBuilder {
  foo: String,
  bar: Option<u32>
}
```

but because I wanted to combine multiple sources into one config. I ended up with:

```rs
#[derive(Configuration)]
struct ConfigBuilder {
  foo: Option<String>,
  bar: Option<u32>
}
```

so now it seems that `required` has changed semantics a little bit

Reply:
You were correct about `config(required)`. I fixed it. Now only `Configuration::build` will fail only if a required field is missing, it will take a default for all other fields

Reply:
> And also, there is no need to use Option<Option<..>> type in ConfigurationBuilder fields. All of the fields might be simple Option<..>, but if the field is marked as #[config(optional)], it is also Option<..> in the Configuration. it changes behaviour during the configuration build (i.e. ConfigurationBuilder → Configuration):

check the code, there should be no more `Option<Option<_>>` in the builder

Reply:
> check the code, there should be no more Option<Option<_>> in the builder

Yeah, I meant that my proposal doesn't require to bring it back.

> > Could we implement defaults binding in the same way serde does it?
> 
> we could, but I maintain that default values should be constant so it doesn't make sense to make it expression

Isn't this the example of simple constant default values?

```rust
#[derive(Configuration)]
struct Config {
  #[config(default_value = "Lorem")]
  foo: String
}
```

Moreover, you can see and edit the default value right in the place of the field definition, and you don't have to lookup for `DEFAULT_FOO` constant somewhere else to navigate too.

Reply:
> In the second option:
> 
>     * You don't have to manually specify `DEFAULT_FOO` - it is generated by macro from `default_value` attr.
> 
>     * You have a default value specified right next (_well, before_) to the field itself
> 
>     * You don't need to specify that the field is required: it is required by definition unless it doesn't have a default value.
> 
>     * You can have a completely optional field. I am not sure that we need it now, but you can painlessly add it later.
> 
>     * You have default fallback in `ConfigBuilder::build()`. Current way to have defaults is to merge `ConfigBuilder::default()` with `ConfigBuilder::from_file()` with `ConfigBuilder::from_env()` and then finally call `.build()`:

I will apply basically everything you have pointed out here